src/main/java/io/fabric8/elasticsearch/discovery/kubernetes/KubernetesUnicastHostsProvider.java

```diff
@@ -76,13 +82,81 @@ public KubernetesUnicastHostsProvider(Settings settings,
   public List<DiscoveryNode> buildDynamicNodes() {
     final List<DiscoveryNode> result = new ArrayList<>();
     AccessController.
-      doPrivileged((PrivilegedAction) () -> {
+      doPrivileged((PrivilegedAction<Void>) () -> {
         result.addAll(readNodes());
         return null;
       });
 
     return result;
   }
+  
+  private List<DiscoveryNode> getDiscoveryNodes(List<Pod> pods, String ipAddress) {
+  	List<DiscoveryNode> discoveryNodes = new ArrayList<>();
+  	
+    pods.stream().forEach(pod -> {
+      String ip = pod.getStatus().getPodIP();
+      try {
+        InetAddress podAddress = InetAddress.getByName(ip);
+        String formattedPodAddress = NetworkAddress.format(podAddress);
+        if (formattedPodAddress.equals(ipAddress)) {
+          // We found the current node.
+          // We can ignore it in the list of DiscoveryNode
+          logger.trace("current node found. Ignoring {}", podAddress);
+        } else {
+          try {
+            TransportAddress[] addresses = transportService.addressesFromString(formattedPodAddress + ":" + this.podPort, 1);
+            for (TransportAddress transportAddress : addresses) {
+              logger.info("adding pod {}, transport_address {}", podAddress, transportAddress);
+              discoveryNodes.add(new DiscoveryNode("#cloud-" + podAddress + "-" + 0, transportAddress, version.minimumCompatibilityVersion()));
+            }
+          } catch (Exception e) {
+            logger.warn("failed to add pod {}", e, podAddress);
+          }
+        }
+      } catch (UnknownHostException e) {
+        logger.warn("Ignoring invalid pod IP address: {}", e, ip);
+      }
+    });
+    return discoveryNodes;
+  }
+
+  private List<DiscoveryNode> getDiscoveryNodes(Endpoints endpoints, String ipAddress) {
+    List<DiscoveryNode> discoveryNodes = new ArrayList<>();
+    endpoints.getSubsets().stream().forEach(endpointSubset -> {
+      endpointSubset.getAddresses().stream().forEach((address -> {
+        String ip = address.getIp();
+        try {
+          InetAddress endpointAddress = InetAddress.getByName(ip);
+          String formattedEndpointAddress = NetworkAddress.format(endpointAddress);
+          try {
+            if (formattedEndpointAddress.equals(ipAddress)) {
+              // We found the current node.
+              // We can ignore it in the list of DiscoveryNode
+              logger.trace("current node found. Ignoring {}", ipAddress);
+            } else {
+              endpointSubset.getPorts().stream().forEach((port) -> {
+                try {
```

Comment:
Looks like there is an opportunity to refactor this block into a private method for reuse here and  the previous method